web/e2e_test.go

```diff
@@ -259,6 +259,369 @@ func checkNeedles(t *testing.T, ts *httptest.Server, req string, needles []strin
+		t.Fatalf("Add: %v", err)
+	}
+	s := searcherForTest(t, b)
+	srv := Server{
+		Searcher: s,
+		Top:      Top,
+		HTML:     true,
+	}
+
+	mux, err := NewMux(&srv)
+	if err != nil {
+		t.Fatalf("NewMux: %v", err)
+	}
+
+	ts := httptest.NewServer(mux)
+	defer ts.Close()
+
+	for req, needles := range map[string][]FileMatch{
+		"/search?q=our&format=json&ctx=0": {
+			{
+				FileName: "f2",
+				Repo:     "name",
+				Matches: []Match{
+					{
+						FileName: "f2",
+						LineNum:  4,
+						Fragments: []Fragment{
+							{
+								Pre:   "f",
+								Match: "our",
+								Post:  "th",
+							},
+						},
+					},
+				},
+			},
+		},
+		"/search?q=our&format=json&ctx=2": {
+			{
+				FileName: "f2",
+				Repo:     "name",
+				Matches: []Match{
+					{
+						FileName: "f2",
+						LineNum:  4,
+						Fragments: []Fragment{
+							{
+								Pre:   "f",
+								Match: "our",
+								Post:  "th",
+							},
+						},
+						Before: []string{
+							"second snippet",
+							"third thing",
+						},
+						After: []string{
+							"fifth block",
+							"sixth example",
+						},
+					},
+				},
+			},
+		},
+		"/search?q=one&format=json&ctx=2": {
+			{
+				FileName: "f2",
+				Repo:     "name",
+				Matches: []Match{
+					{
+						FileName: "f2",
+						LineNum:  1,
+						Fragments: []Fragment{
+							{
+								Pre:   "",
+								Match: "one",
+								Post:  " line",
+							},
+						},
+						After: []string{
+							"second snippet",
+							"third thing",
+						},
+					},
+				},
+			},
+		},
+		"/search?q=seventh&format=json&ctx=2": {
+			{
+				FileName: "f2",
+				Repo:     "name",
+				Matches: []Match{
+					{
+						FileName: "f2",
+						LineNum:  7,
+						Fragments: []Fragment{
+							{
+								Pre:   "",
+								Match: "seventh",
+								Post:  "",
+							},
+						},
+						Before: []string{
+							"fifth block",
+							"sixth example",
+						},
+					},
+				},
+			},
+		},
+		"/search?q=seventh&format=json&ctx=10": {
+			{
+				FileName: "f2",
+				Repo:     "name",
+				Matches: []Match{
+					{
+						FileName: "f2",
+						LineNum:  7,
+						Fragments: []Fragment{
+							{
+								Pre:   "",
+								Match: "seventh",
+								Post:  "",
+							},
+						},
+						Before: []string{
+							"one line",
+							"second snippet",
+							"third thing",
+							"fourth",
+							"fifth block",
+							"sixth example",
+						},
+					},
+				},
+			},
+		},
+		"/search?q=one&format=json&ctx=10": {
+			{
+				FileName: "f2",
+				Repo:     "name",
+				Matches: []Match{
+					{
+						FileName: "f2",
+						LineNum:  1,
+						Fragments: []Fragment{
+							{
+								Pre:   "",
+								Match: "one",
+								Post:  " line",
+							},
+						},
+						After: []string{
+							"second snippet",
+							"third thing",
+							"fourth",
+							"fifth block",
+							"sixth example",
+							"seventh",
+						},
+					},
+				},
+			},
+		},
+	} {
+		checkResultMatches(t, ts, req, needles)
+	}
```

Comment:
Thank you for adding so many tests!

Three more test cases please:

- filename matches with context
- non-zero ctx with a file which ends with a newline.
- non-zero ctx and empty lines in the context

Reply:
Added tests. Though one thing you'll notice is that I'm not swallowing the final `\n` in `Before`, `After` and `Fragments` (we already don't return the one in the fragments anyway).

So if I have ctx=1 `\nfoo\n<match>\nbar\n` then the result will be `before=foo` and after=`bar`.